File: python/linked_list/reverse_linked_list.py
"""
Problem: Reverse Linked List
Given the head of a singly linked list, reverse the list, and return the reversed list.

Approaches:
1. Iterative pointer reversal O(n)
2. Recursive O(n) + call stack
"""
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # TODO: iterative
        # TODO: recursive
        raise NotImplementedError

if __name__ == "__main__":
    # manual small test
    n3 = ListNode(3)
    n2 = ListNode(2, n3)
    n1 = ListNode(1, n2)
    s = Solution()
    rev = s.reverseList(n1)
    while rev:
        print(rev.val, end=" ")
        rev = rev.next
    print()

File: cpp/linked_list/reverse_linked_list.cpp
/*
Problem: Reverse Linked List
Approaches:
1. Iterative pointer reversal O(n)
2. Recursive O(n)
*/
#include <bits/stdc++.h>
using namespace std;

struct ListNode {
    int val; ListNode* next; ListNode(int v): val(v), next(nullptr) {}
};

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        // TODO: iterative then recursive alt
        return head; // placeholder
    }
};

int main() {
    ListNode* n1 = new ListNode(1);
    n1->next = new ListNode(2);
    n1->next->next = new ListNode(3);
    Solution s;
    ListNode* r = s.reverseList(n1);
    while (r) { cout << r->val << ' '; r = r->next; }
    cout << endl;
}
